### PR TITLE
Add support for Snapshots and first unit tests

### PR DESCRIPTION
**Pull request by lexming**
Changelog:
* add methods `list_snapshots`, `create_filesystem_snapshot` and `delete_filesystem_snapshot` with the same interface as in `GpfsOperations`
*  `create_filesystem_snapshot` and `delete_filesystem_snapshot`  support snapshots of dtrees, but the names of the snapshots are automatically prefixed with the name of the dtree. This is a limitation of OceanStor, which uses a single namespace for snapshots for the whole filesystem.
* add unit test for the new snapshot methods and also all underlying methods used by them, such as `list_filesystems` and `get_filesystem_info`
* add mock object to the tests that mimics the REST API of OceanStor

### CODE REVIEW

**Comment by lexming** [comment]
Tests fail because the pipeline cannot fetch `vsc-config`

File: lib/vsc/filesystem/oceanstor.py [hunk 1]
```diff
@@ -1810,3 +1810,141 @@ def _get_grace_expiration(grace, usage, soft, hard):
...
+
+        fs = self.get_filesystem_info(filesystem)
+        filter_json = {"file_system_id": int(fs["id"])}
+
+        if fileset is not None:
+            dtree = self.get_fileset_info(filesystem, fileset)
+            if dtree is None:
+                err_msg = "Snapshot query failed: fileset '%s' not found in filesystem '%s'" % (filesystem, fileset)
+                self.log.raiseException(err_msg, OceanStorOperationError)
+            filter_json.update({"dtree_id": dtree["id"]})
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: any particular reason you're not just adding it in the dict directly? `filter_json['dtree_od'] = dtree['id']`?

File: lib/vsc/filesystem/oceanstor.py [hunk 2]
```diff
@@ -1810,3 +1810,141 @@ def _get_grace_expiration(grace, usage, soft, hard):
...
+                filesets = [filesets]
+
+            for fileset in filesets:
+                if self.get_fileset_info(fsname, fileset) is None:
+                    self.log.error("Cannot create snapshot: fileset %s not found on filesystem %s!", fileset, fsname)
+                    return 0
+
+                # the snapshot namespace of all filesets is shared in OceanStor
+                # create unique snapshot name per fileset
+                fileset_snapname = "%s_%s" % (fileset, snapname)
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: I would definitely add debug print that you're doing this?

File: lib/vsc/filesystem/oceanstor.py [hunk 3]
```diff
@@ -1810,3 +1810,141 @@ def _get_grace_expiration(grace, usage, soft, hard):
...
+        Create new snapshot in OceanStor
+
+        @type snap_name: string representing the name of the new snapshot
+        @type fs_name: name of the filesystem of the new snapshot
+        @type fileset_name: name of the dtree fileset of the new snapshot
+        """
+        snapshots = self.list_snapshots(fs_name, fileset_name)
+        if snap_name in snapshots:
+            self.log.error("Snapshot '%s' already exists for filesystem %s!", snap_name, fs_name)
+            return 0
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py:1879]:
```suggestion
            return False
```

File: lib/vsc/filesystem/oceanstor.py [hunk 4]
```diff
@@ -1810,3 +1810,141 @@ def _get_grace_expiration(grace, usage, soft, hard):
...
+                # the snapshot namespace of all filesets is shared in OceanStor
+                # create unique snapshot name per fileset
+                fileset_snapname = "%s_%s" % (fileset, snapname)
+
+                new_snap_status = self._new_snapshot_api(fileset_snapname, fsname, fileset)
+        else:
+            # filesystem snapshot
+            new_snap_status = self._new_snapshot_api(snapname, fsname, None)
+
+        return new_snap_status
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py:1866]: maybe we should return the list of created snapshots?

File: lib/vsc/filesystem/oceanstor.py (hunk 1 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: nothing in particular, changed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/24/commits/71038c50ba8a005f08da15dcb614474a54f1a9a4

**Comment by wpoely86** [comment]
Can you add a file like: https://github.com/vub-hpc/eb_hooks/blob/main/vsc-ci.ini and rerun `python -m vsc.install.ci`?

File: lib/vsc/filesystem/oceanstor.py (hunk 2 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: good point, I also made a new private method to make this name change. Fixed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/24/commits/3a01e6dd3325f8252670d28c9f21a389a89f06b8

File: lib/vsc/filesystem/oceanstor.py (hunk 3 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py:1879]: I don't like this either, but this is done to match what `GpfsOperations.create_filesystem_snapshot` returns.

File: lib/vsc/filesystem/oceanstor.py (hunk 4 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py:1866]: This mimics the behavior of `GpfsOperations.create_filesystem_snapshot`. Alternatively, we can have the list of snapshots in a class attribute and update it before this return. Would that work for you?

**Review by wpoely86 — APPROVED** [review]

**Review by wpoely86 — APPROVED** [review]

**Comment by wpoely86** [comment]
There are issues with the pipeline but let's that not hold us back for now.

**Pull request merged**